### PR TITLE
Laboratory work #4, Nikolas Zvarichuk - 23FPL1

### PR DESCRIPTION
**Pull request by Niko-Zvar**

### CODE REVIEW

File: lab_4_retrieval_w_clustering/main.py [hunk 1]
```diff
@@ -40,6 +42,10 @@ def __init__(self) -> None:
         """
         Initialize an instance of the BM25Vectorizer class.
         """
+        Vectorizer.__init__(self, [])
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:45]: use `super`

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

**Review by artyomtugaryov — COMMENTED** [review]
Good, But I'm waiting 10

File: lab_4_retrieval_w_clustering/main.py (hunk 1 above)
Comment (Niko-Zvar) [reply lab_4_retrieval_w_clustering/main.py:48]: done

File: lab_4_retrieval_w_clustering/main.py [hunk 2]
```diff
@@ -3,9 +3,16 @@
 
 Vector search with clusterization
 """
+from lab_2_retrieval_w_bm25.main import calculate_bm25
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: all imports together!

File: lab_4_retrieval_w_clustering/main.py [hunk 3]
```diff
@@ -40,6 +50,9 @@ def __init__(self) -> None:
         """
         Initialize an instance of the BM25Vectorizer class.
         """
+        self._corpus = []
+        self._avg_doc_len = -1.0
+        super().__init__(self._corpus)
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: super should be first line in init

File: lab_4_retrieval_w_clustering/main.py [hunk 4]
```diff
@@ -81,6 +109,15 @@ def _calculate_bm25(self, tokenized_document: list[str]) -> Vector:
         Returns:
             Vector: BM25 vector for document.
         """
+        vector_bm_25 = [0.0] * len(self._vocabulary)
+
+        bm_25 = calculate_bm25(self._vocabulary, tokenized_document,
+                        self._idf_values, 1.5, 0.75,
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: Do not set `1.5, 0.75` - it is default

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
I'm waiting 10!

File: lab_4_retrieval_w_clustering/main.py (hunk 2 above)
Comment (Niko-Zvar) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 3 above)
Comment (Niko-Zvar) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 4 above)
Comment (Niko-Zvar) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/main.py [hunk 5]
```diff
@@ -51,6 +64,15 @@ def set_tokenized_corpus(self, tokenized_corpus: TokenizedCorpus) -> None:
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not tokenized_corpus:
+            raise ValueError("Empty tokenized_corpus")
+        if self._corpus:
+            self._corpus = []
+        summary_len = 0
+        for token in tokenized_corpus:
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: Looks to hard

File: lab_4_retrieval_w_clustering/main.py [hunk 6]
```diff
@@ -113,6 +154,27 @@ def put_corpus(self, corpus: Corpus) -> None:
...
+            tokens = self._tokenizer.tokenize(text)
+            if not isinstance(tokens, list):
+                raise ValueError('Tokens is not list')
+            tokenized_corpus.append(tokens)
+        if None in tokenized_corpus:
+            raise ValueError('NoneType in tokenized_corpus')
+        self._vectorizer.set_tokenized_corpus(tokenized_corpus)
+        for i, tokenized_text in enumerate(tokenized_corpus):
+            if not tokenized_text:
+                tokenized_corpus.pop(i)
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:170]: why?

File: lab_4_retrieval_w_clustering/main.py [hunk 7]
```diff
@@ -170,6 +252,8 @@ def __init__(self, db: DocumentVectorDB) -> None:
         Args:
             db (DocumentVectorDB): Object of DocumentVectorDB class.
         """
+        self._db = db
+        super().__init__(self._db.get_vectorizer(), self._db.get_tokenizer())
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: should be the first line in init

File: lab_4_retrieval_w_clustering/main.py [hunk 8]
```diff
@@ -182,6 +266,23 @@ def retrieve_relevant_documents(self, query: str, n_neighbours: int) -> list[tup
...
+        if not tokenized_query:
+            raise ValueError('Empty tokenized_query')
+        query_vector = self._db.get_vectorizer().vectorize(tokenized_query)
+        if not query_vector:
+            raise ValueError('Empty query_vector')
+        self._db.put_corpus(self._db.get_raw_documents())
+        docs_vectors = [vec[1] for vec in self._db.get_vectors()]
+        relevant_documents = self._calculate_knn(query_vector, docs_vectors, n_neighbours)
+        if relevant_documents is None:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: Add message

File: lab_4_retrieval_w_clustering/main.py [hunk 9]
```diff
@@ -287,6 +411,30 @@ def run_single_train_iteration(self) -> list[ClusterDTO]:
...
+                distances.append(distance)
+            self.__clusters[distances.index(min(distances))].add_document_index(i)
+        for cluster in self.__clusters:
+            cl_vectors = [vectors[index][1] for index in cluster.get_indices()]
+            new_centroid = []
+            for vector in zip(*cl_vectors):
+                value = sum(vector) / len(cl_vectors)
+                new_centroid.append(value)
+            cluster.set_new_centroid(tuple(new_centroid))
+        return self.__clusters
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: the method must return NEW! Not set this clusters, but return new

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_4_retrieval_w_clustering/main.py (hunk 9 above)
Comment (Niko-Zvar) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 8 above)
Comment (Niko-Zvar) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 7 above)
Comment (Niko-Zvar) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 5 above)
Comment (Niko-Zvar) [reply lab_4_retrieval_w_clustering/main.py]: changed

File: lab_4_retrieval_w_clustering/main.py (hunk 6 above)
Comment (Niko-Zvar) [reply lab_4_retrieval_w_clustering/main.py:170]: because of test (connected with stop words)

**Pull request closed without merging**